### PR TITLE
Changing Courses Retrieval Method in BottomBar component

### PR DESCRIPTION
**Pull request by rohanmaheshwari430**
### Summary 

[Courses Interface and Migration Design Doc](https://docs.google.com/document/d/1dLwbJuufBTilqnJ3XinqMrvYeR6x8Zf17LvglcxjWhU/edit)

In this PR, we change the procedure in which the BottomBar component retrieves course information. We also correct the logic regarding which semester's course information we want to fetch.

What this PR actually does: 

_Before:_ 

When a course card is clicked, the information for that course is retrieved from the class roster API. The roster semester that is passed in this request is pulled from the user's copy of the course in 'user-semesters' Firestore collection.

_After:_ 
1. Course card is clicked.
2. Pass down the season and year variables to the BottomBar component.
3. Season and year variables are turned into the semester format for class roster API (i.e Spring 2023 -> SP23). 
4. Attempt to fetch course information for that semester (via 'courses' collection).
5. If successful, add it to the BottomBar.
6. If unsuccessful, fetch the latest roster available for that course (via 'available-rosters-for-course' collection).
7. Fetch the course information for latest semester (via 'courses' collection, which will be guaranteed success).
8. Add it to BottomBar.

This pull request is the second step towards implementing the Courses backend into the Frontend.
- [x] BottomBar component now references the courses firestore.
- [x] Handled cases when the course offering does not align with the semester the user put it in (i.e JAPAN 1101 is only offered in the fall, but the user has it in Spring 2023). 
- [x] Implemented some data conversion helper functions between CoursePlan's semester format and class roster's semester format (i.e Spring 2023 -> SP23 and vice versa)

Remaining TODOs:

- [ ] Change the VuexStoreState to use our new proposed Courses Interface (details can be found in design doc)

### Test Plan 

TBD

### Notes 

Need to do a thorough check of courses info usage in CoursePlan before we get rid of any previous courses collections in Firestore.

### Breaking Changes  

- Added Firestore usage rules for 'courses' and 'available-rosters-for-course' collections

### CODE REVIEW

**Comment by noschiff** [comment]
haven't reviewed but SUPER EXCITING!!!

**Comment by zachary0kent** [comment]
Also, it seems like the accessibility tests for the bottom bar are failing.

**Review by zachary0kent — CHANGES_REQUESTED** [review]
The type errors in the unchanged files are due to the fact that additional required properties were added to the `course` component, which are not passed into the `course` components used in the template of `IncompleteSubReqCourse.vue`

**Comment by rohanmaheshwari430** [comment]
> The type errors in the unchanged files are due to the fact that additional required properties were added to the `course` component, which are not passed into the `course` components used in the template of `IncompleteSubReqCourse.vue`

Wow this makes a lot of sense. Will definitely fix this up.

File: src/global-firestore-data/courses.ts [hunk 1]
```diff
@@ -81,21 +83,6 @@ const getCourse = async (
...
-const seasonAndYearToRosterIdentifier = (season: FirestoreSemesterSeason, year: number): string => {
-  const seasonToSemesterMap = {
-    Fall: 'FA',
-    Spring: 'SP',
-    Winter: 'WI',
-    Summer: 'SU',
-  } as const;
-
-  return `${seasonToSemesterMap[season]}${year - 2000}`;
-};
```
Comment (zachary0kent) [inline src/global-firestore-data/courses.ts:97]: Why was this removed?

Comment (zachary0kent) [reply src/global-firestore-data/courses.ts:97]: I think this implementation is better than the existing one

Comment (rohanmaheshwari430) [reply src/global-firestore-data/courses.ts:97]: Gotcha, i updated the implementation, but still removed the duplicate one.

File: src/components/Course/Course.vue [hunk 2]
```diff
@@ -81,6 +81,8 @@ export default defineComponent({
     active: { type: Boolean, required: true },
     isReqCourse: { type: Boolean, required: true },
     semesterIndex: { type: Number, required: false, default: 0 },
+    season: { type: String, required: false, default: '' },
+    year: { type: Number, required: false, default: 0 },
```
Comment (zachary0kent) [inline src/components/Course/Course.vue:85]: I think making these properties optional to fix the type error is kind of sketchy. Is there a reason why they might not be needed sometimes?

**Review by zachary0kent — CHANGES_REQUESTED** [review]
Just a couple more nits

File: src/components/Course/Course.vue (hunk 2 above)
Comment (rohanmaheshwari430) [reply src/components/Course/Course.vue:85]: Because in the implementation of retrieving a course, if a valid season and year is not provided for the course, we will retrieve the latest offering of that course. So by having a default value, we are pretty much saying that we want the latest offering of the course.

File: src/components/BottomBar/BottomBarState.ts [hunk 3]
```diff
@@ -81,11 +93,27 @@ export const addCourseToBottomBar = (course: FirestoreSemesterCourse): void => {
...
+  } else {
+    // if not, retrieve the info from the latest class roster the course is in
+    const latestRosterIndex = availableRostersForCourse.rosters.length - 1;
+    const latestRoster = availableRostersForCourse.rosters[latestRosterIndex];
+    const seasonAndYear = rosterIdentifierToSeasonAndYear(latestRoster);
+    vueForBottomBar.bottomCourses = [
+      firestoreSemesterCourseToBottomBarCourse(course, seasonAndYear.season, seasonAndYear.year),
+      ...vueForBottomBar.bottomCourses,
+    ];
+  }
```
[the remark this answers is not in the document]
Comment (rohanmaheshwari430) [reply src/components/BottomBar/BottomBarState.ts:115]: Yessir

File: src/components/BottomBar/BottomBarState.ts [hunk 4]
```diff
@@ -100,21 +128,23 @@ export const addCourseToBottomBar = (course: FirestoreSemesterCourse): void => {
...
+      const bottomBarCourse = vueForBottomBar.bottomCourses.find(
+        ({ uniqueID, code }) => uniqueID === course.uniqueID && code === course.code
+      );
+      if (bottomBarCourse) {
+        bottomBarCourse.description = description;
+        bottomBarCourse.prereqs = prereqs;
+        bottomBarCourse.enrollment = enrollment;
+        bottomBarCourse.lectureTimes = lectureTimes;
+        bottomBarCourse.instructors = instructors;
+        bottomBarCourse.distributions = distributions;
```
[the remark this answers is not in the document]
Comment (rohanmaheshwari430) [reply src/components/BottomBar/BottomBarState.ts:141]: Yessir

**Review by zachary0kent — APPROVED** [review]
Looks good!
